File: Data Structures & Algorithms/surrounded-regions/submission-4.py
class Solution:
    def solve(self, board: List[List[str]]) -> None:
        n,m = len(board), len(board[0])
        q = []
        # vis = [[False for _ in range(m)] for _ in range(n)]
        for i in range(n):
            if board[i][0] == 'O':
                q.append((i,0))
                board[i][0] = 'T'
            if board[i][m-1] == 'O':
                q.append((i,m-1))
                board[i][m-1] = 'T'

        for i in range(m):
            if board[0][i] == 'O':
                q.append((0,i))
                board[0][i] = 'T'
            if board[n-1][i] == 'O':
                q.append((n-1,i))
                board[n-1][i] = 'T'
        
        delr = [-1, 0, 1, 0]
        delc = [0, -1, 0, 1]

        while q:
            x,y = q.pop(0)
            for i in range(4):
                nrow, ncol = x + delr[i], y + delc[i]
                if 0 <= nrow < n and 0 <= ncol < m and board[nrow][ncol] == 'O':
                    board[nrow][ncol] = 'T'
                    q.append((nrow, ncol))

        for i in range(n):
            for j in range(m):
                if board[i][j] == 'O':
                    board[i][j] = 'X'
        for i in range(n):
            for j in range(m):
                if board[i][j] == 'T':
                    board[i][j] = 'O'
            
        

File: Data Structures & Algorithms/search-for-word/submission-4.cpp
class Solution {
public:
    bool solve(int x, int y, string word, vector<vector<char>> &board) {
        if (word == "") return true;
        if (x < 0 || y < 0 || x >= board.size() || y >= board[0].size() || board[x][y] == '#' || board[x][y] != word[0]) {
            return false;
        }
        char og = board[x][y];
        board[x][y] = '#';
        bool res = solve(x+1, y, word.substr(1,word.size()), board) || 
                    solve(x, y + 1, word.substr(1, word.size()), board) ||
                    solve(x-1, y, word.substr(1,word.size()), board) || 
                    solve(x, y-1, word.substr(1,word.size()), board);
        board[x][y] = og;
        return res;
    }
    bool exist(vector<vector<char>>& board, string word) {
        for (int i = 0; i < board.size(); i++) {
            for (int j = 0; j < board[0].size(); j++) {
                if (board[i][j] == word[0]) {
                    if (solve (i, j, word, board)) return true;
                }
            }
        }
        return false;
    }
};
